homeassistant/requirements.py

```diff
@@ -151,48 +162,45 @@ async def async_process_requirements(
     """
     if (pip_lock := hass.data.get(DATA_PIP_LOCK)) is None:
         pip_lock = hass.data[DATA_PIP_LOCK] = asyncio.Lock()
-    install_failure_history = hass.data.get(DATA_INSTALL_FAILURE_HISTORY)
-    if install_failure_history is None:
-        install_failure_history = hass.data[DATA_INSTALL_FAILURE_HISTORY] = set()
-
-    kwargs = pip_kwargs(hass.config.config_dir)
+    install_failure_history = hass.data.setdefault(DATA_INSTALL_FAILURE_HISTORY, set())
+    is_installed_cache = hass.data.setdefault(DATA_IS_INSTALLED_CACHE, set())
+    missing = [req for req in requirements if req not in is_installed_cache]
+    if not missing:
+        return
+    for req in missing:
+        if req in install_failure_history:
+            _LOGGER.info(
+                "Multiple attempts to install %s failed, install will be retried after next configuration check or restart",
+                req,
+            )
+            raise RequirementsNotFound(name, [req])
 
     async with pip_lock:
-        for req in requirements:
-            await _async_process_requirements(
-                hass, name, req, install_failure_history, kwargs
-            )
+        await _async_process_requirements(
+            hass,
+            name,
+            missing,
+            is_installed_cache,
+            install_failure_history,
+        )
 
 
 async def _async_process_requirements(
     hass: HomeAssistant,
     name: str,
-    req: str,
+    requirements: list[str],
+    is_installed_cache: set[str],
     install_failure_history: set[str],
-    kwargs: Any,
 ) -> None:
     """Install a requirement and save failures."""
-    if req in install_failure_history:
-        _LOGGER.info(
-            "Multiple attempts to install %s failed, install will be retried after next configuration check or restart",
-            req,
-        )
+    kwargs = pip_kwargs(hass.config.config_dir)
+    for req in requirements:
+        if await hass.async_add_executor_job(_install_if_missing, req, kwargs):
+            is_installed_cache.add(req)
+            continue
+        install_failure_history.add(req)
         raise RequirementsNotFound(name, [req])
```

Comment:
Since we call multiple times into the executor, better to just run this whole method in an executor so we don't bounce multiple times between event loop and executor.

Reply:
44c5d646f39afa4f5269e93d194402b508b75cce